lib/codegen/CreateTerm.cpp

```diff
@@ -199,15 +205,53 @@ llvm::Value *getBlockHeader(llvm::Module *Module, KOREDefinition *definition, co
+static llvm::Value *addrspaceCast(llvm::Value *val, llvm::BasicBlock *block, int from, int to) {
+  auto AllocType = val->getType()->getPointerElementType();
+  bool hasUnnamed = false;
+  std::string name = "addrspace_" + std::to_string(from) + "_to_" + std::to_string(to) + "_" + getMangledTypeStr(AllocType, hasUnnamed);
+  auto addrspace = llvm::CallInst::Create(getOrInsertFunction(block->getModule(), name, llvm::PointerType::get(AllocType, to), llvm::PointerType::get(AllocType, from)), {val}, "", block);
+  return addrspace;
+}
+
+llvm::Value *addrspaceCast0to1(llvm::Value *val, llvm::BasicBlock *block) {
+  return addrspaceCast(val, block, 0, 1);
+}
+
+llvm::Value *addrspaceCast1to0(llvm::Value *val, llvm::BasicBlock *block) {
+  return addrspaceCast(val, block, 1, 0);
+}
+
+llvm::Value *ptrToInt(llvm::Value *val, llvm::BasicBlock *block) {
+  bool hasUnnamed = false;
+  std::string name = "ptrtoint_i64." + getMangledTypeStr(val->getType(), hasUnnamed);
+  auto ptr = llvm::CallInst::Create(getOrInsertFunction(block->getModule(), name, llvm::Type::getInt64Ty(block->getModule()->getContext()), val->getType()), {val}, "", block);
+  return ptr;
 }
 
-llvm::Value *allocateTerm(llvm::Type *AllocType, llvm::Value *Len, llvm::BasicBlock *block, const char *allocFn) {
-  llvm::Instruction *Malloc = llvm::CallInst::CreateMalloc(block, llvm::Type::getInt64Ty(block->getContext()), AllocType, Len, nullptr, koreHeapAlloc(allocFn, block->getModule()));
-  setDebugLoc(&block->getInstList().back());
-  block->getInstList().push_back(Malloc);
-  return Malloc;
+llvm::Value *allocateTerm(ValueType Cat, llvm::Type *AllocType, llvm::BasicBlock *block, const char *allocFn) {
+  return allocateTerm(Cat, AllocType, llvm::ConstantExpr::getSizeOf(AllocType), block, allocFn);
+}
+
+llvm::Value *allocateTerm(ValueType Cat, llvm::Type *AllocType, llvm::Value *Len, llvm::BasicBlock *block, const char *allocFn) {
+  auto malloc = llvm::CallInst::Create(koreHeapAlloc(Cat, allocFn, block->getModule()), Len, "malloccall", block);
+  if (malloc->getType()->getPointerElementType() == AllocType) {
+    return malloc;
+  }
+  auto cast = new llvm::BitCastInst(malloc, llvm::PointerType::get(AllocType, 1), "", block);
+  return cast;
+}
+
+llvm::Value *allocateTermNoReloc(llvm::Type *AllocType, llvm::BasicBlock *block, const char *allocFn) {
+  return allocateTermNoReloc(AllocType, llvm::ConstantExpr::getSizeOf(AllocType), block, allocFn);
+}
+
+llvm::Value *allocateTermNoReloc(llvm::Type *AllocType, llvm::Value *Len, llvm::BasicBlock *block, const char *allocFn) {
+  auto malloc = llvm::CallInst::Create(koreHeapAlloc(allocFn, block->getModule()), llvm::ConstantExpr::getPtrToInt(llvm::ConstantExpr::getGetElementPtr(AllocType, llvm::ConstantPointerNull::get(llvm::PointerType::get(AllocType, 1)), llvm::ConstantInt::get(llvm::Type::getInt32Ty(block->getContext()), 1)), llvm::Type::getInt64Ty(block->getContext())), "malloccall", block);
```

Comment:
Only a stylistic issue, but I find this quite difficult to read as the line is so long.